web-app/HeimScripts/heatmap/run.R

```diff
@@ -54,31 +59,85 @@ main <- function(max_rows = 100, sorting = "nodes") {
   write(jsn,file = "heatmap.json")   # json file be served the same way
                                      # like any other file would - get name via
                                      # /status call and then /download
-  
+
   msgs <- c("Finished successfuly")
   list(messages = msgs)
 }
 
-addStats <- function(df, sorting) {
+writeMarkerTable <- function(markerTable){
+  colnames(markerTable) <- c("rowLabel", "biomarker",
+                             "log2FoldChange", "t", "pValue", "adjustedPValue", "B")
+  jsn                   <- toJSON(markerTable, pretty = TRUE)
+  write(jsn, file = markerTableJson)
+}
+
+getMeasurements <- function(df) {
+  if (ncol(df) > 3){
+    return(df[, 3:ncol(df)])
+  }
+  else {
+    return( df[3] )
+  }
+}
+
+
+cleanUpLimmaOutput <- function() {
+  if (file.exists(markerTableJson)) {
+    file.remove(markerTableJson)
+  }
+}
+
+hasTwoSubsets <- function(measurements) {
+  getSubset1Length(measurements) < ncol(measurements)
+}
+
+addStats <- function(df, sorting, ranking, max_rows) {
+  measurements  <- getMeasurements(df)
+  rankingMethod <- getRankingMethod(ranking)
+  twoSubsets    <- hasTwoSubsets(measurements)
   if (ncol(df) > 3) {
     #this is the case for more than 1 sample
     df <-
-      applySorting(df,sorting)  # no need for sorting in one sample case
-    variances <-
-      apply(df[,3:ncol(df)],1,var, na.rm = T)  # Calculating
-    # variance per probe (per row)
-    means <- rowMeans(df[,3:ncol(df)], na.rm = T)  # this is just an
-                                                   # auxiliary column - it will not be
-                                                   # used for JSON.
-    sdses <- apply(df[,3:ncol(df)],1,sd, na.rm = T)  # this is just
-                                                     # an auxiliary column -
-                                                     # it will not be used for JSON.
-    df["MEAN"] <- means
-    df["SD"] <- sdses
-    df["SIGNIFICANCE"] <- variances
-    df <- df[with(df, order(-SIGNIFICANCE)),]
+      applySorting(df,sorting)  # no need for sorting in one sample case, we do it here only
+    if ( twoSubsets ){
```

Comment:
Marker table is always produced and displayed if there are 2 subsets?
Is that in the requirements, or should the user have the option to select that they want to run statistics for finding differentially expressed genes? (Does this for example take a lot of time for big data.frames? In that case it might be better to make it optional?)